### PR TITLE
findMinVertexModel() cleaned

### PR DESCRIPTION
**Pull request by ombhilare999**
image

image

- In findMinVertexModel(y):
```
for all g nodes:
    //------------------ Routing Setup ---------------------//
    ripUpRouting(y, G);         //Ripup the previous routing
    (*Users)[i].push_back(y);   //Users update                 
    invUsers[y] = i;            //InvUsers update
    //------------------------------------------------------//

   //Attempts placement for the given y(HID) using:
  totalCosts[i] += routeSignal(G, H, y, gConfig); //This cost is based on out-going edges of Y.

  For all of the input of the y:
      //------------------ Routing Setup ---------------------//
      ripUpRouting(driverHNode, G);
      (*Users)[driverGNode].push_back(driverHNode); //Users update
      invUsers[driverHNode] = driverGNode;          //InvUsers update
      //------------------------------------------------------//

      // Newfeature: rip up from load: ripUpLoad(G, driver, outputPin);
      totalCosts[i] += routeSignal(G, H, driverHNode, gConfig);

  //Track the best FunCell based on total cost (Routing out-going edges of Y as well incoming edges of Y)

//For the best FunCell: ripUpRouting last time and route the signal once-again.
```

- In routeSignal(): Routes all outgoing edges of the application node and measure the total cost for it.
```
For all out-going edges of signal-HID:

    //----------------------------------------------------------------//
    //--------- Finding appropriate outputPin for the driver ---------//
    //----------------------------------------------------------------//

    //------------------------------------------------------//
    //    Adding outputPin in routing data structure:       //
    //(This is used as a starting point in route() function)//
    //------------------------------------------------------//
    (*Users)[driverOutPin].push_back(y);
    (*Trees)[y].nodes.push_back(driverOutPin);
     //------------------------------------------------------//


    //----------------------------------------------------------------//

    //----------------------------------------------------------------//
    //--------- Finding appropriate inputPin for the load ------------//
    //----------------------------------------------------------------// 
     //No need to add inputPin to the routing datastructure, it is shared as the target in the route() function.

    //Invoke route function which is simple Dijkstra's algorithm:
   cost = route(G, y, loadInPin, &path, gConfig);  //loadInPin is the hit location.
```

### CODE REVIEW

**Comment by ombhilare999** [comment]
Hi @HamasWaqar ,
Can you review this PR and merge it?
Thank you.

**Comment by HamasWaqar** [comment]
> image
> 
> image
> 
> * In findMinVertexModel(y):
> 
> ```
> for all g nodes:
>     //------------------ Routing Setup ---------------------//
>     ripUpRouting(y, G);         //Ripup the previous routing
>     (*Users)[i].push_back(y);   //Users update                 
>     invUsers[y] = i;            //InvUsers update
>     //------------------------------------------------------//
> 
>    //Attempts placement for the given y(HID) using:
>   totalCosts[i] += routeSignal(G, H, y, gConfig); //This cost is based on out-going edges of Y.
> 
>   For all of the input of the y:
>       //------------------ Routing Setup ---------------------//
>       ripUpRouting(driverHNode, G);
>       (*Users)[driverGNode].push_back(driverHNode); //Users update
>       invUsers[driverHNode] = driverGNode;          //InvUsers update
>       //------------------------------------------------------//
> 
>       // Newfeature: rip up from load: ripUpLoad(G, driver, outputPin);
>       totalCosts[i] += routeSignal(G, H, driverHNode, gConfig);
> 
>   //Track the best FunCell based on total cost (Routing out-going edges of Y as well incoming edges of Y)
> 
> //For the best FunCell: ripUpRouting last time and route the signal once-again.
> ```
> 
>  * In routeSignal(): Routes all outgoing edges of the application node and measure the total cost for it.
> 
> ```
> For all out-going edges of signal-HID:
> 
>     //----------------------------------------------------------------//
>     //--------- Finding appropriate outputPin for the driver ---------//
>     //----------------------------------------------------------------//
> 
>     //------------------------------------------------------//
>     //    Adding outputPin in routing data structure:       //
>     //(This is used as a starting point in route() function)//
>     //------------------------------------------------------//
>     (*Users)[driverOutPin].push_back(y);
>     (*Trees)[y].nodes.push_back(driverOutPin);
>      //------------------------------------------------------//
> 
> 
>     //----------------------------------------------------------------//
> 
>     //----------------------------------------------------------------//
>     //--------- Finding appropriate inputPin for the load ------------//
>     //----------------------------------------------------------------// 
>      //No need to add inputPin to the routing datastructure, it is shared as the target in the route() function.
> 
>     //Invoke route function which is simple Dijkstra's algorithm:
>    cost = route(G, y, loadInPin, &path, gConfig);  //loadInPin is the hit location.
> ```

1. What is "apt" stand for again? I may be forgetting the full form of the word
2. Based on our earlier discussion, I have a vague understanding of the new RouteSignal function. Still trying to wrap my head around it. However, could you elaborate for documentation purposes on how this one function can find the placement for both the driver and load? More so
- how does this function vary if we are placing either the load or driver pin
- In the code, RouteSignal function is called multiple times in findMinVertex - could you elaborate on the purpose of each RouteSignal function
3. First Code Block:  Is the `HID` in `y(HID)` for the driver or load node in the application graph
`//Attempts placement for the given y(HID) using:
  totalCosts[i] += routeSignal(G, H, y, gConfig); //This cost is based on out-going edges of Y.`
4. First code Block: In our code, we use y as a term for driver, but good to document this for others since the code is now public?
5. Second Code Block: for documentation purposes, is the definition of y in RouteSignal function different from the definition in findMinVertex?
6. Second Code Block: It mentions at the end of the code that there is no need for input pins in the RouteSignal function. However, routeSignal function is being called multiple times from the minVertex model for placing and routing both the driver and load pin. I do not get why their is no need for input pins routing in this code block. Maybe I am getting confused on it, so best to document this right now.

File: src/UGRAMM.cpp [hunk 1]
```diff
@@ -58,22 +58,23 @@ int findMinVertexModel(DirectedGraph *G, DirectedGraph *H, int y, std::map<int,
   boost::tie(ei, ei_end) = in_edges(yD, *H);
   for (; ei != ei_end; ei++)
   {
-    if ((*Trees)[source(*ei, *G)].nodes.size())
+    //if ((*Trees)[source(*ei, *G)].nodes.size())
+    if (invUsers[source(*ei, *G)] != NOT_PLACED)
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp]: **The following image is from GRAMM research paper:** 
image

**and its implementation is as follows:**
https://github.com/UniversalGRAMM/UGRAMM/blob/26880ba43eea79cd620cafffded7765f9d46384e/src/UGRAMM.cpp#L51C4-L98

**In summary:**
- The above part of the code, checks whether adjacent cells (Input and Output) of current application-node have any vertex model defined or not. (Vertex model is nothing but pin-to-pin mapping information; If vertex model is defined for a node, then it MUST Be placed which is indicated by the invUsers data-structure.)

File: src/UGRAMM.cpp [hunk 2]
```diff
@@ -58,22 +58,23 @@ int findMinVertexModel(DirectedGraph *G, DirectedGraph *H, int y, std::map<int,
...
       allEmpty = false;
   }
 
   out_edge_iterator eo, eo_end;
   boost::tie(eo, eo_end) = out_edges(yD, *H);
   for (; eo != eo_end; eo++)
   {
-    if ((*Trees)[target(*eo, *H)].nodes.size())
+    //if ((*Trees)[target(*eo, *H)].nodes.size())
+    if (invUsers[target(*eo, *H)] != NOT_PLACED)
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp]: **The following image is from GRAMM research paper:** 
image

**and its implementation is as follows:**
https://github.com/UniversalGRAMM/UGRAMM/blob/26880ba43eea79cd620cafffded7765f9d46384e/src/UGRAMM.cpp#L51C4-L98

**In summary:**
- The above part of the code, checks whether adjacent cells (Input and Output) of current application-node have any vertex model defined or not. (Vertex model is nothing but pin-to-pin mapping information; If vertex model is defined for a node, then it MUST Be placed which is indicated by the invUsers data-structure.)

File: src/UGRAMM.cpp [hunk 3]
```diff
@@ -82,28 +83,22 @@ int findMinVertexModel(DirectedGraph *G, DirectedGraph *H, int y, std::map<int,
...
       if (compatibilityCheck((*gConfig)[chooseRand].Type, (*hConfig)[y].Opcode))
       {
-        //  Finding the output Pin for the selected FuncCell:
-        selectedCellOutputPin = findOutputPinFromFuncCell(chooseRand, G);
         vertex_found = true;
       }
     }
 
-    UGRAMM->debug("[RandomSelection] for application node :: {} :: Choosing starting vertex model as :: {} :: {}", hNames[y], gNames[chooseRand], gNames[selectedCellOutputPin]);
+    UGRAMM->debug("[RandomSelection] for application node :: {} :: Choosing starting vertex model as :: {}", hNames[y], gNames[chooseRand]);
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp:88]: - The above part of the code, checks whether adjacent cells (Input and Output) of current application-node have any vertex model defined or not. (Vertex model is nothing but pin-to-pin mapping information; If vertex model is defined for a node, then it MUST Be placed which is indicated by the invUsers data-structure.)
**- If neither input nor output nodes of the node y is not placed/mapped, a random (suitable---same compatibility type) node is returned.**

File: src/UGRAMM.cpp [hunk 4]
```diff
@@ -122,26 +117,22 @@ int findMinVertexModel(DirectedGraph *G, DirectedGraph *H, int y, std::map<int,
...
+    //------------------ Routing Setup ---------------------//
+    ripUpRouting(y, G);         //Ripup the previous routing
+    (*Users)[i].push_back(y);   //Users update                 
+    invUsers[y] = i;            //InvUsers update
+    //------------------------------------------------------//
 
     // Cost and history costs are calculated for the FuncCell:
     totalCosts[i] += (1 + (*HistoryCosts)[i]) * ((*Users)[i].size() * PFac);
 
+    //Placement of the signal Y:
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp:127]: https://github.com/UniversalGRAMM/UGRAMM/blob/26880ba43eea79cd620cafffded7765f9d46384e/lib/UGRAMM.h#L151

- Y is the application node for which we are finding minimum vertex model.
- As shown in this example from GRAMM paper:
image

File: src/UGRAMM.cpp [hunk 5]
```diff
@@ -239,22 +215,20 @@ int findMinVertexModel(DirectedGraph *G, DirectedGraph *H, int y, std::map<int,
...
-    (*Users)[driverPinNode].push_back(driver);
-    invUsers[driver] = driverFuncCellNode;
+    //------------------ Routing Setup ---------------------//
+    ripUpRouting(driverHNode, G);
+    (*Users)[driverGNode].push_back(driverHNode); //Users update
+    invUsers[driverHNode] = driverGNode;          //InvUsers update
+    //------------------------------------------------------//
 
-    routeSignal(G, H, driver, gConfig);
+    routeSignal(G, H, driverHNode, gConfig);
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp:229]: https://github.com/UniversalGRAMM/UGRAMM/blob/396efc291fc5718655212b515e8442db8b9873bc/src/UGRAMM.cpp#L201

https://github.com/UniversalGRAMM/UGRAMM/blob/396efc291fc5718655212b515e8442db8b9873bc/src/UGRAMM.cpp#L212

- We have ripUp any previous connections before attempting to route for the bestIndexCell

File: src/routing.cpp [hunk 6]
```diff
@@ -418,54 +418,110 @@ int routeSignal(DirectedGraph *G, DirectedGraph *H, int y, std::map<int, NodeCon
...
+
     std::string H_LoadPin = boost::get(&EdgeProperty::H_LoadPin, *H, *eo);
     std::string H_DriverPin = boost::get(&EdgeProperty::H_DriverPin, *H, *eo);
 
     if (load == y)
       continue; // JANDERS ignore feedback connections for the moment
 
-    if(invUsers[load] == -1)
+    //UserCheck for load:
+    if(invUsers[load] == NOT_PLACED)
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/routing.cpp:450]: > So, is the placement of the load being done in the routing function?
- The placement of the load has to been done in previous iterations of findMinVertex();
- Here as the earlier comments suggested, this is nothing but a safety check. To be able to use the inputPin of the load in the routing path, the load application node at first place must be placed in the device model.

https://github.com/UniversalGRAMM/UGRAMM/blob/26880ba43eea79cd620cafffded7765f9d46384e/src/routing.cpp#L428-L456

**Comment by ombhilare999** [comment]
> What is "apt" stand for again? I may be forgetting the full form of the word
- Appropriate (apt) driver selection depending on user preferences. If the user enters driver=PinA on the Edge, the output pin of the related FunCell named PinA is selected.

---

> how does this function vary if we are placing either the load or driver pin
- We don't need to place the "Pins"

---

> In the code, RouteSignal function is called multiple times in findMinVertex - could you elaborate on the purpose of each RouteSignal function

- In the findMinVertex function, we initially place the current application node (HID/Y) by using the routeSignal function. The routeSignal function calculates the routing cost based on the outgoing edges of node Y. Afterward, for every incoming edge of node Y, we call routeSignal again. This means that when routeSignal(Driver) is executed, it returns the routing cost for the outgoing edges of the driver node, i.e., the cost of routing the driver to the current node Y. Ultimately, for node Y, the costs of both incoming and outgoing edges are summed up, and the best cumulative cost is selected after evaluating all possible device model positions.

image

- findMinVertex(y); //The application node y for which we are finding the minmum vertex
  - For all suitable G nodes:
       -  Placement of node y first @ g (outgoing edge cost):
            -   cost += routeSignal(y);
            -  The above function will route the outgoing edges of y, as shown in the above image. It will route and return the cost of y->z.
       - Incoming edges of y:
            -  cost += routeSignal(x1); //Cost of x1->y
            -  cost += routeSignal(x2); //Cost of x1->y
            -  cost += routeSignal(x3); //Cost of x1->y
    - The total cost of incoming and outgoing edge is noted and after trying to place a node y at all suitable device model nodes, best device model node is found out based on the cost.

File: src/UGRAMM.cpp [hunk 7]
```diff
@@ -155,27 +146,24 @@ int findMinVertexModel(DirectedGraph *G, DirectedGraph *H, int y, std::map<int,
...
-      invUsers[driver] = driverFuncCellNode;
+      //------------------ Routing Setup ---------------------//
+      ripUpRouting(driverHNode, G);
+      (*Users)[driverGNode].push_back(driverHNode); //Users update
+      invUsers[driverHNode] = driverGNode;          //InvUsers update
+      //------------------------------------------------------//
 
       // Newfeature: rip up from load: ripUpLoad(G, driver, outputPin);
-      totalCosts[i] += routeSignal(G, H, driver, gConfig);
+      totalCosts[i] += routeSignal(G, H, driverHNode, gConfig);
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp:161]: Check: https://github.com/UniversalGRAMM/UGRAMM/pull/16#issuecomment-2377911990

**Pull request merged**